### ISSUE TITLE
Monstache Stops Syncing After Repeated HTTP 504 Errors

### ISSUE THREAD

**Issue by patrickserrano** [rwynn/monstache#570]
Hi @rwynn

My team is in the process of deploying Monstache to our staging environment and we're noticing some strange behavior that I was hoping you might have some insight on.

We're seeing instances where Monstache is getting HTTP 504 errors over a period of time, and then it just stops syncing data until we've restarted the Monstache processes. We're using Elasticsearch v7.10.1 with Mongo v4.0.12 and Monstache v6.7.5

I'm working with our DevOps team to try and figure out why we're seeing the 504 errors in the first place, but I didn't see anything in the Monstache documentation that would suggest that a string of timeouts like that would kill the processes. But that _seems_ to be the behavior that we're seeing at the moment.

We're deploying Monstache in Kubernetes with 2 pods so that we have high availability with this config:

```toml
direct-read-namespaces = [
    "collection-1",
    "collection-2",
    "collection-3",
    "collection-4"
]

change-stream-namespaces = [
    "collection-1",
    "collection-2",
    "collection-3",
    "collection-4"
]

mapper-plugin-path="/lib/plugin.so"
exit-after-direct-reads = false
direct-read-stateful = true
resume = true
resume-write-unsafe = true
resume-strategy = 0
post-processors = 10
elasticsearch-retry = true
elasticsearch-max-bytes = 16000000
elasticsearch-max-seconds = 10
enable-easy-json = true
gzip = true
enable-http-server = true
http-server-addr = ":8080"
stats = true
index-stats = true
verbose = true
```

Thanks a ton for any help!

**Comment by rwynn** [comment]
Hi @patrickserrano ,

If the monstache process is dying it may mean it is taking too much memory and getting killed.  You have a pretty high `elasticsearch-max-bytes` value.  That is not really a problem, but when monstache is not able to successfully flush those bulk actions, and more actions pile in from MongoDB it can eventually run out of memory.  

The problem is that monstache uses the `BulkProcessorService` from https://github.com/olivere/elastic and that doesn't currently call `wp.service.Reset()` [here](https://github.com/olivere/elastic/blob/release-branch.v7/bulk_processor.go#L580) when the error is non-nil.  So the backlog of requests keeps growing.  

In some sense this is good for transient errors that work themselves out quickly since you don't lose data.  But if the failures persist for too long that request queue can get too big (well over the `elasticsearch-max-bytes`.  

If my hunch is right you may be able to mitigate a little by setting `elasticsearch-max-bytes` much lower (default is 8mb).

**Comment by patrickserrano** [comment]
@rwynn thanks for such a quick response!

I'll take a look at tweaking the `elasticsearch-max-byte` value since our stg and prod clusters will have more nodes than our dev cluster did. 

With that said, looking at the pod resource usage over the last 7 days, I'm not seeing the RAM consumption getting anywhere near what's been allocated for Monstache. Each pod had been allocated 4GB but we are consistently under 1GB of memory usage. Is there a limit on how much memory the go processes can consume that maybe I need to take a look at increasing?

**Comment by rwynn** [comment]
Is the process dying or just seems to be doing nothing?  There is a failure mode it can go into where it is waiting for a healthy response from Elasticsearch.

https://github.com/olivere/elastic/blob/e037a96c4d3a384002c071cb999b0f32e35e2dd5/bulk_processor.go#L508

**Comment by patrickserrano** [comment]
@rwynn we're seeing logs like this: `elastic: <our-es-url>  is dead`

Sometimes when we get that log Monstache doesn't resume syncing until we kill the process. Other times, like this morning, there's a period of time (1 hour this morning) and then monstache will start syncing change again. 

We did change `elasticsearch-max-bytes`and `elasticsearch-max-seconds` back to the default values which has greatly improved the overall stability of the process. But the intermittent `elasticsearch is dead` errors are vexing since we can't find any pattern to them.

As an aside, I do have another question for you. When `exit-after-direct-read` is enabled, does the exist occur after the direct read only is completed, or after the direct read + changes that happened since the direct read started? I remember reading in another issue that you mentioned monstache stores a timestamp when the direct reads are started so that it knows where to start reading changes from, but I'm unclear if _any_ changes are synced when that option is enabled. Thanks again!

**Comment by rwynn** [comment]
@patrickserrano I think the `is dead` message comes when there is a timeout running `HEAD` request against the configured Elastic URL.  

Monstache will start shutdown as soon as all the direct reads have completed.  It might pick up a few change events until the go routines have completely stopped (received that message).  If you have configured change streams those events happen while direct reads are occurring.  

The `direct-read-bounded` setting, if enabled, takes into account the first and last document ID in the collection before starting the direct read.  So it will not read past the last document, even if something was inserted in the middle of the read. This insert might still be indexed if you have change streams enabled though.  It just means that it will not be via direct reads.

**Comment by patrickserrano** [comment]
@rwynn thanks for all of your help, I think we got this sorted out. I think the 504s were caused by setting the `elasticsearch-max-bytes` and `elasticsearch-max-seconds` values too hight. We reverted to the default and the issues look to have cleared up. We have seen the occasional `elasticsearch is dead` error log, but those seem to correlate with nodes cycling with a deployment.

I'm going to close this issue. 

Also thanks for the info on the direct-reads and change streams, it's greatly informed some design decisions for our post-MVP deployment design.
